### PR TITLE
Testing subcommand (and other changes to testing)

### PR DESCRIPTION
**Pull request by xmo-odoo**
## Test subcommand

Adds a test subcommand through py.test (with ~all py.test options forwarded directly):
- flexible testing specification, the subcommand can take Odoo modules, directories and/or files as positional parameters and will apply [py.test's discovery conventions](http://pytest.org/latest/goodpractises.html#test-discovery) to fetch Python test cases, "data" test cases will be extracted from a module's `__openerp__.py` if that matches other filters
- _only runs selected tests_, if requesting the testing of `product` all its dependencies (and auto_install dependents) are installed (with demo data) but only the tests of `product` are run
- with no positional parameters, just starts looking for tests in the current working directory
- automatically creates and deletes a temporary db if none is provided, if one is provided does not delete it (but creates it if it's missing) which can allow reusing dbs
- uses an inited (but not specifically configured) Odoo logger and a pytest in default display mode (line per test file, captures stdout/stderr)
- recap at the end of testing with display of failed test stack traces & captured stderr/stdout
- provided marks to ignore categories of tests, only current mark is `http` so `odoo.py test -m 'not http'` will not start an http server and won't run any HTTP test
- existing "unittest" tests mostly work as-is (unless they were mucking with the registry or even runtest protocol), alternatively can write lighter "pytest" tests
- fixtures make for pretty clean setup/teardown & composition
- slightly to greatly faster tests depending on the selected module(s)
- future improvements via existing addons (e.g. [pep8](https://pypi.python.org/pypi/pytest-pep8), [coverage](https://pypi.python.org/pypi/pytest-cover)) or more formal results reporting (e.g. [junitxml result logs](http://pytest.org/latest/usage.html#creating-junitxml-format-files)) or new addons for e.g. better JS or tour integration

e.g. `odoo.py test mail account` (will run all non-skipped tests in the mail and account addons), `odoo.py test  addons/product/tests/test_pricelist.py` (will just run the two test cases in test_pricelist.py), `odoo.py test --pdb -k test_20_pricelist_uom` (will only run that test case, should drop into pdb if the test fails)
## `--test-enable`
- most of the registry-installation functions moved to methods of Registry
- Registry.load_modules changed to a coroutine and testing part moved outside of it to RegistryManager.new, which external runner does _not_ use see 3219863
- "on-install" testing uses same test discovery subcommand does with custom test reporter
  - reporting not identical to old reporting but pretty close (and improvement in some ways)
  - replaced old assertion_reporter (which just counted) by normal assertion errors in XML and YAML
  - all failed tests summarised/printed at the end of installation, no need to go hunt in the log
- remains compatible with runbot (I think, went into runbot source to see how that worked)
## other changes
- some tests were implicitly skipped (weren't imported by test package), changed to explicit skipping
- various imports improved, some of them may not be necessary after all. May be useful if we eventually try to make just calling `py.test` directly possible
- moved and removed imports so pytest and unittest2 are only imported when actually trying to _run_ tests
- removed `--test-file` behaviour, superseded by `odoo.py test <testfile>`

/cc @odony @antonylesuisse @tde-banana-odoo

### CODE REVIEW

**Comment by KangOl** [comment]
Can you rebase our branch?

**Comment by xmo-odoo** [comment]
?

Rebase on the current master?

**Comment by KangOl** [comment]
Yes, because "This branch has conflicts that must be resolved"

**Comment by xmo-odoo** [comment]
> Yes, because "This branch has conflicts that must be resolved"

Rebased on master, tests pass.

**Comment by Yenthe666** [comment]
Guys is there any ETA for this? This looks interesting! :+1:

File: doc/reference/testing.rst [hunk 1]
```diff
@@ -23,38 +22,34 @@ e.g.
         |-- test_bar.py
         `-- test_foo.py
 
-and ``__init__.py`` contains::
+.. versionchanged:: 9.0
```
Comment (KangOl) [inline doc/reference/testing.rst]: 10.0

File: openerp/api.py [hunk 2]
```diff
@@ -862,7 +873,10 @@ def __new__(cls, cr, uid, context):
 
         # otherwise create environment, and add it in the set
         self = object.__new__(cls)
-        self.cr, self.uid, self.context = self.args = (cr, uid, frozendict(context))
+        self.args = (cr, uid, frozendict(context))
```
Comment (KangOl) [inline openerp/api.py]: Now,
```python
self.args[2] is not self.context
```

Is it a expected change?

File: openerp/tools/config.py [hunk 3]
```diff
@@ -387,7 +385,7 @@ def die(cond, msg):
             'debug_mode', 'dev_mode', 'smtp_ssl', 'load_language',
             'stop_after_init', 'logrotate', 'without_demo', 'xmlrpc', 'syslog',
             'list_db', 'proxy_mode',
-            'test_file', 'test_enable', 'test_commit', 'test_report_directory',
+            'test_file', 'test_enable', 'test_report_directory',
```
Comment (KangOl) [inline openerp/tools/config.py]: `test_file` shouldn't be also removed?

File: openerp/api.py (hunk 2 above)
Comment (xmo-odoo) [inline openerp/api.py]: Wouldn't think so, though it shouldn't make any difference given it's a frozendict.

File: openerp/tools/config.py (hunk 3 above)
Comment (xmo-odoo) [inline openerp/tools/config.py]: Probably should, yes.

**Comment by bringsvor** [comment]
Excellent idea. Sure would like to use this.

**Comment by Fenkiou** [comment]
Any news on this one ? :)

This seems really interesting as for now, when installing a module with test enable, it runs the tests for all module (in case of database init) or I'm doing it wrong.

**Comment by tebanep** [comment]
Is this still on the roadmap?

**Comment by tde-banana-odoo** [comment]
Any news on this ?

**Pull request closed without merging**